src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+                - 'log': Silently replace errors with `null` and
+                    log the errors in :attr:`errors`.
+
+            errorfunc: A function with signature `f(x, e) -> Any`,
+                where `x` are the original column values as a :class:`pd.Series` indexed
+                by the source input dataframe names, and `e` is the original error.
+                When `errors='log'`, the returned value is logged instead of the
+                original error.
+
+        Returns:
+            A dictionary of dataframes, harvested and aggregated (if `aggregate`)
+            from the input dataframes, matching the name and schema of the resources.
+            The index are resource key fields if `aggregate`=`True`,
+            and the name of the input dataframe otherwise.
+
+        Raises:
+            ValueError: Resource has duplicate field basenames.
+            KeyError: Missing columns needed by resource.
+        """
+        resources = self.prune(resources)
+        self.test(resources)
+        odfs = {}
+        reports = {}
+        for resource in resources:
+            rname = resource["name"]
+            harvest = resource.get("harvest", False)
+            columns = [field["name"] for field in resource["schema"]["fields"]]
+            key = resource["schema"]["primaryKey"]
+            samples = {}
+            if harvest:
+                for name, df in self.dfs.items():
+                    sdf = sample(df, key=key, data=columns)
+                    if sdf is not None:
+                        samples[name] = sdf
+            else:
+                samples[rname] = self.dfs[rname][columns]
+            # Pass source dataframe names to aggregate via the index
+            for name, sdf in samples.items():
+                samples[name] = sdf.set_index(pd.Index([name] * len(sdf), name="df"))
+            odf = pd.concat(samples.values())[columns]
+            if aggregate:
+                if harvest:
+                    aggfuncs = {
+                        field["name"]: field["aggregate"]
+                        for field in resource["schema"]["fields"]
+                        if field["name"] not in key
+                    }
+                    if errors == "log":
+                        odf, report = groupby_aggregate(
+                            odf,
+                            by=key,
+                            aggfuncs=aggfuncs,
+                            errors="report",
+                            errorfunc=errorfunc,
+                        )
+                        if report:
+                            reports[rname] = report
+                    else:
+                        odf = groupby_aggregate(
+                            odf,
+                            by=key,
+                            aggfuncs=aggfuncs,
+                            errors=errors,
+                            errorfunc=errorfunc,
+                        )
```

Comment:
On your first point, good catch. I've simplified the code.
Logging errors is slower and more memory intensive than coercing errors to `null`. But if you do not see a use case for `errors='coerce'` (i.e. you always want either a full error report or to fail on first error with `errors='raise'`), I'm fine with scrapping it.

Reply:
If it is much quicker without logging then that makes sense to me!

Reply:
`errors='log'` being slower must have been in the past, because based on new speed tests, it is now just as fast as `errors='coerce'`. Shall I?

- Drop `ResourceBuilder.build(..., errors='coerce')`
- Rename `log` to `report`.
- Drop `ResourceBuilder.report()` and make `ResourceBuilder.build()` always return both the build result and the report. If `errors='raise'`, the report can be `{}`.

Reply:
If I follow what you are suggesting that sounds good.

Basically you'd only ever raise errors or report them. and if there are no errors, when you are trying to raise errors the report should be empty. But the method will always return some report even if it is empty.